### PR TITLE
Refactor operations

### PR DESCRIPTION
**Pull request by vinnamkim**
### Summary
 - This file has too many lines (2007 lines) in it: https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/components/operations.py 
 - Move annotation matchers and mergers to the seperate directory
 - Move get_merger to the seperate directory (it is a pre-work for refactoring mergers.)

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [x] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Review by cih9088 — DISMISSED** [review]

File: datumaro/components/annotations/matcher.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        b_segm = b_segms[matched_b]
+
+        if label_matcher(a_segm, b_segm):
+            matches.append((a_segm, b_segm))
+        else:
+            mispred.append((a_segm, b_segm))
+
+    # *_umatched: boxes of (*) we failed to match
+    a_unmatched = [a_segms[i] for i, m in enumerate(a_matches) if m < 0]
+    b_unmatched = [b_segms[i] for i, m in enumerate(b_matches) if m < 0]
```
Comment (JihwanEom) [inline datumaro/components/annotations/matcher.py]: `b_segms` is not changed in for statement. Do we need `if len(b_segms) == 0: break` in for statement?

**Review by sooahleex — DISMISSED** [review]
Looks good to me.

**Review by wonjuleee — CHANGES_REQUESTED** [review]
Why did you introduce annotations directory and what stands for? We may have prior understandings to make the same consensus. Could you explain?

**Review by wonjuleee — COMMENTED** [review]
Are you thinking about introducing merge directory as a python API refactoring manner?

**Comment by vinnamkim** [comment]
> Why did you introduce annotations directory and what stands for? We may have prior understandings to make the same consensus. Could you explain?

First of all, this file has too many lines (2007 lines) in it: https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/components/operations.py. Therefore, many classes and functions for different purposes are defined in this file. This is the full list of definitions in this file:
image

My motivation is to refactor merge related things to somewhere but not in `operations.py`.
To achieve it, I firstly try to refactor `operations.py` for `Matcher`s and `Merger`s.

File: datumaro/components/annotations/matcher.py (hunk 1 above)
Comment (vinnamkim) [reply datumaro/components/annotations/matcher.py]: Fixed by https://github.com/openvinotoolkit/datumaro/pull/901/commits/b3f72d24005794a1d13a5998045a9d0803d6b5a7

**Comment by wonjuleee** [comment]
> > Why did you introduce annotations directory and what stands for? We may have prior understandings to make the same consensus. Could you explain?
> 
> First of all, this file has too many lines (2007 lines) in it: https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/components/operations.py. Therefore, many classes and functions for different purposes are defined in this file. This is the full list of definitions in this file: image
> 
> My motivation is to refactor merge related things to somewhere but not in `operations.py`. To achieve it, I firstly try to refactor `operations.py` for `Matcher`s and `Merger`s.

I agree with separating Matcher and Merger from operations. But my question is about annotation directory. I would like to know what annotation directory means for Matcher and Merger.

**Review by bonhunko — APPROVED** [review]
LGTM

**Comment by vinnamkim** [comment]
> I agree with separating Matcher and Merger from operations. But my question is about annotation directory. I would like to know what annotation directory means for Matcher and Merger.

Since `Matcher` and `Merger` correspond to each `Annotation` 1:1, it seems appropriate move them under `annotations` directory. So ultimately, all `Annotation`s should be moved to `annotations` directory as well.

**Comment by wonjuleee** [comment]
> > I agree with separating Matcher and Merger from operations. But my question is about annotation directory. I would like to know what annotation directory means for Matcher and Merger.
> 
> Since `Matcher` and `Merger` correspond to each `Annotation` 1:1, it seems appropriate move them under `annotations` directory. So ultimately, all `Annotation`s should be moved to `annotations` directory as well.

alright, I agree that matcher should be closely related to the annotation-level operation. But I think that merger is more about dataset-level operation. What do you think? Do you have any plan to create dataset directory in a similar vein?
What about merge directory? Could you share your plan or opinion about refactoring components in Datumaro?

**Comment by vinnamkim** [comment]
> But I think that merger is more about dataset-level operation. What do you think? Do you have any plan to create dataset directory in a similar vein? What about merge directory? Could you share your plan or opinion about refactoring components in Datumaro?

1. `Merger`s corresponding to each `Annotation` (e.g., `LabelMerger`) which are in charge of element-wise merge operations are moved to `./datumaro/components/annotations` directory.
2. Merge policies (for dataset level, e.g., `ExactMerge` or `IntersectMerge`) are moved to `./datumaro/components/merge` directory.

My plan was like this, and in this PR, 1) is completed and 2) is partially completed.

**Review by wonjuleee — APPROVED** [review]

**Pull request merged**